pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -753,4 +748,117 @@ public IdealState apply(IdealState idealState) {
+    final ZNRecord partitionAssignment = getKafkaPartitionAssignment(realtimeTableName);
+    final Map<String, List<String>> partitionToServersMap = partitionAssignment.getListFields();
+    final KafkaStreamMetadata kafkaStreamMetadata = new KafkaStreamMetadata(tableConfig.getIndexingConfig().getStreamConfigs());
+
+    final String realtimeServerTenantName =
+        ControllerTenantNameBuilder.getRealtimeTenantNameForTenant(tableConfig.getTenantConfig().getServer());
+    final List<String> currentInstances = getInstances(realtimeServerTenantName);
+
+    // Previous partition count is what we find in the Kafka partition assignment znode.
+    // Get the current partition count from Kafka.
+    final int prevPartitionCount = partitionToServersMap.size();
+    int currentPartitionCount = -1;
+    try {
+      currentPartitionCount = getKafkaPartitionCount(kafkaStreamMetadata);
+    } catch (Exception e) {
+      LOGGER.warn("Could not get partition count for {}. Leaving kafka partition count at {}", realtimeTableName, currentPartitionCount);
+      return;
+    }
+
+    // Previous instance set is what we find in the Kafka partition assignment znode (values of the map entries)
+    final Set<String> prevInstances = new HashSet<>(currentInstances.size());
+    for (List<String> servers : partitionToServersMap.values()) {
+      prevInstances.addAll(servers);
+    }
+
+    final int prevReplicaCount = partitionToServersMap.entrySet().iterator().next().getValue().size();
+    final int currentReplicaCount = Integer.valueOf(tableConfig.getValidationConfig().getReplicasPerPartition());
+
+    boolean updateKafkaAssignment = false;
+
+    if (!prevInstances.equals(new HashSet<String>(currentInstances))) {
+      LOGGER.info("Detected change in instances for table {}", realtimeTableName);
+      updateKafkaAssignment = true;
+    }
+
+    if (prevPartitionCount != currentPartitionCount) {
+      LOGGER.info("Detected change in Kafka partition count for table {} from {} to {}", realtimeTableName, prevPartitionCount, currentPartitionCount);
+      updateKafkaAssignment = true;
+    }
+
+    if (prevReplicaCount != currentReplicaCount) {
+      LOGGER.info("Detected change in per-partition replica count for table {} from {} to {}", realtimeTableName, prevReplicaCount, currentReplicaCount);
+      updateKafkaAssignment = true;
+    }
+
+    if (!updateKafkaAssignment) {
+      LOGGER.info("Not updating Kafka partition assignment for table {}");
+      return;
+    }
+
+    // Generate new kafka partition assignment and update the znode
+    if (currentInstances.size() < currentReplicaCount) {
+      LOGGER.error("Cannot have {} replicas in {} instances for {}.Not updating partition assignment", currentReplicaCount, currentInstances.size(), realtimeTableName);
+      return;
+    }
+    ZNRecord newPartitionAssignment = generatePartitionAssignment(kafkaStreamMetadata.getKafkaTopicName(), currentPartitionCount, currentInstances, currentReplicaCount);
```

Comment:
I'd log the fact that we're updating the partition assignment and/or that it completed.